### ISSUE TITLE
Revise lighting model for foam

### ISSUE THREAD

**Issue by huwb** [huwb/crest-oceanrender#21]
This issue is to track improvement of lighting on the foam. It currently looks flat right now. Some kind of 3D lighting would probably improve it a lot.

From Midda during conversation on #4 :

> I don't know if this is a bit out-of-scope, but I was thinking about why the foam looks too saturated at sunset/sunrise. After just looking at some sunset beach images, I think it might be related to the entire foam surface being uniformly affected. Because real foam has some height, you get a bit of shading on it.

> [Sunset foam example 1](http://images4.fanpop.com/image/photos/19900000/SunSet-sunsets-and-sunrises-19955107-2560-1817.jpg)
> [Sunset foam example 2](https://2.bp.blogspot.com/-erAA6Gno9mY/TleruNnAgGI/AAAAAAAATg0/yvYrt0d31aU/s1600/Beach-sunset-wallpaper.jpg)

> You get bright saturated colours on the sides facing the light, but shadows/ambient on the opposite side. Maybe a normal map (and some PBR roughness) for the foam might help give some more believable shading?

**Comment by huwb** [comment]
@Midda-C had some success today. I just pushed some stuff to a new branch foam_lighting, commit ed6c4f13fe5ce9e2e508a08104f6cfaf612dce33

I'm generating normals for the foam that give it curved edges. It looks like this:

![image](https://user-images.githubusercontent.com/939901/40878727-073a465e-668d-11e8-8bd3-dcc78488a01e.png)

What do you think?

While working on this i realised the bubble/underwater foam also doesnt work well at all, mainly because its just slapped onto the surface. I've now pushed it under the surface by adding refraction offsets, dialling down the motion a bit so it moves independently of the surface white foam, and added a bit of parallax based on view direction to push the texture downwards under the surface. I really like the result, its a million times better IMO. Needs to be seen under motion. Let me know what you think if you can get this branch, otherwise I'll get it promoted to master if you're happy with the above image (as an initial pass at least) and after i've done a bit of cleanup.

**Comment by huwb** [comment]
One more shot for good measure:

![image](https://user-images.githubusercontent.com/939901/40878896-c37c12c8-668f-11e8-8552-19355ff06404.png)

**Comment by Midda-C** [comment]
Yeah, just tried it out, it's looking good!  Probably only thing I think is left really are some properties to adjust the thresholds for where foam appears (e.g. height/steepness, opacity, etc), but that's not really related to the lighting/shading.

Nice work!

**Comment by huwb** [comment]
Thanks dude. I did a bunch of cleanup and refactoring today and cleared up some of the mess that accumulated over time in the foam code.

I've added a toggle to turn 3D foam shading on and off. I've added 3 parameters for controlling where foam appears - the ones beginning with "Wave Foam ..". I also added two parameters to tweak the 3D foam shading - a directional light scale and a normal strength.

This is all pushed to master now. I'll close this as i believe the feature dev part is done (now for the bugs!). Let me know how you get on with it.

**Comment by huwb** [comment]
@Midda-C 

The latest merge into master (of the sim branch) brings with it a big rebuild of the foam shading. I removed some ugly stuff and tried to make the foam look much less severe. The foam now adopts the normal of the water before doing NdL (so it should sit better in the frame/scene), and there's a configurable specular component as well as i realised foam should be shiny :). The below is a first pass of tweaking it. It's more perfect but it already feels much more harmonious and natural to me.

![image](https://user-images.githubusercontent.com/939901/43368470-6e3577f8-9355-11e8-9ef2-06642a709cef.png)

As always, apologies for any retweaking this requires, but i hope the results are worth it. Feedback welcome!
